### ISSUE TITLE
Slay dem code dragons

### ISSUE THREAD

**Issue by meeDamian** [eternitywall/javascript-opentimestamps#1]
I suggest on starting with:

```
$ npm i -g xo # or yarn global add xo
$ cd javascript-opentimestamps
$ xo --esnext --space --fix
```

This ↑ will fix many problems already, point out even more and save a lot of time later ;).

Also, just out of curiosity, is there any reason why it's done in an OO way?

**Comment by lvaccaro** [comment]
Thanks, I appreciate your suggestion. JavaScript happiness is really cool :)
We use Map object ECMAScript 2015 (ES6) to store the object crypto-operations. Moreover, we use hierarchical model object in other parts of code. I think ES6 it is more elegant solution instead a custom approach.
From all this we decided to develop in OO style.

**Comment by RCasatta** [comment]
Just committed auto fixes from xo

**Comment by RCasatta** [comment]
Just committed manual fixes to be more compliant with eslint.
From 318 errors to 11.
@meeDamian any pull request to solve the remaining errors are welcomed. 
In particular I am not sure:
* How to deal with overloading in javascript `no-dupe-class-members`
* How to handle the `no-use-before-define` in notary.js
* How to handle the `no-constant-condition` in while with break in context.js and op.js
* How to handle the `guard-for-in` rule

**Comment by meeDamian** [comment]
As for `guard-for-in` it is recommended to wrap content of the iterating loop in: `{}.hasOwnProperty.call(object, property)` ([more](http://eslint.org/docs/rules/guard-for-in))

ex:

```js
// open-timestamps.js:75
for (const calendarUrl in calendarUrls) {
  if ({}.hasOwnProperty.call(calendarUrls, calendarUrl)) {
    this.submitAsync(calendarUrl, timestamp.msg);
  }
}

// context.js:135
for (const x in value) {
  if ({}.hasOwnProperty.call(value, x)) {
    this.writeByte(x);
  }
}
```

As for `no-dupe-class-members`, AFAIK js classes don't allow for method overload ([more](http://eslint.org/docs/rules/no-dupe-class-members)), so instead of defining methods with two different signatures try:

```js
// context.js
class StreamDeserializationContext {

  // …

  readBytes(expectedLength = undefined) {
    if (expectedLength === undefined) {
      expectedLength = this.read_varuint();
    }
    return this.read(expectedLength);
  }

  // …
}
```

As for `no-constant-condition` it is recommended to avoid `while (const)`. Consider using `do…while` or replacing const with the actual condition. Alternatively, if not possible/viable add rule exception, but please make sure to add a comment explaining why ([more](http://eslint.org/docs/rules/no-constant-condition)).

As of `no-use-before-define`, I really don't know because I never use JS in OO way. It might have something to do with the modules are loaded or how child/parent methods should be called. But again, I can't help here :) ([more](http://eslint.org/docs/rules/no-use-before-define)).

## PR: https://github.com/eternitywall/javascript-opentimestamps/pull/2 

**NOTE:** I'm not sure if logic left in `readVarbytes` is correct.

**Comment by RCasatta** [comment]
PR tested and merged, thanks.
We are going to rewrite the while condition for the `no-constant-condition`.
Regarding `no-use-before-define` we haven't a solution so I'll write a comment

**Comment by RCasatta** [comment]
`no-constant-condition` fix 606803e0424a1cb3da7b847e1d7e1aca979902fa

**Comment by RCasatta** [comment]
`no-use-before-define` comment 4914e1b92c811b8ec097d89e865193b0240de389
